core/shared/src/main/scala/zio/Config.scala

```diff
@@ -473,6 +496,125 @@ object Config {
 
   def duration(name: String): Config[zio.Duration] = duration.nested(name)
 
+  def enumeration[D] = new PartiallyAppliedEnumeration[D]
+
+  class PartiallyAppliedEnumeration[D] {
+    def apply[X <: D](
+      desc1: Config[X]
+    )(implicit tag: ClassTag[X]): Config[D] =
```

Comment:
I don't think we need these class tags since we aren't doing anything with them. Also `Config` is covariant so we should be able to simplify this to just using overloaded variants of `enumeration` instead of partial type application.

Reply:
@adamgfraser  Ofcourse! `zio-config` previously was invariant because it had to write back the config as well, and `ClassTag` was required then. The `enumeration` in this draft PR was a blind copy paste of previous implementation.

This is removed in zio-config PR as well. https://github.com/zio/zio-config/pull/1035/files

Thanks for pointing it out!

Note that I managed to get all the indexing behaviour localised in zio-config. So I won't be following up on this PR for the first 4.x release of zio-config.

@jdegoes  did mention to have two versions, one to maintain the backward compatibility and the other having indexing behaviour. But at this stage, I would be more comfortable releasing a version of zio-config first before porting some behaviour back here.